### PR TITLE
Fixed release order

### PR DESCRIPTION
**Pull request by apankratovantonp**
### Details:
 - `ov::Objects` can be released using `object = {}` statement

### Tickets:
 - 73867

### CODE REVIEW

**Comment by apankratovantonp** [comment]
Tested by removing `EXPECT_NO_CRASH` macro

File: src/core/include/openvino/core/any.hpp [hunk 1]
```diff
@@ -347,6 +358,11 @@ class OPENVINO_API Any {
      */
     Any() = default;
 
+    /**
+     * @brief Destructor presereves unload order of implementation object and reference to library
```
Comment (ilya-lavrenov) [inline src/core/include/openvino/core/any.hpp:362]: presereves => preserves

File: src/core/include/openvino/runtime/tensor.hpp [hunk 2]
```diff
@@ -36,23 +36,28 @@ class VariableState;
...
+    Tensor(const std::shared_ptr<InferenceEngine::Blob>& impl, const std::shared_ptr<void>& so);
 
     friend class ov::runtime::Core;
     friend class ov::runtime::InferRequest;
     friend class ov::runtime::RemoteContext;
     friend class ov::runtime::VariableState;
 
 public:
+    /**
+     * @brief Destructor presereves unload order of implementation object and reference to library
```
Comment (ilya-lavrenov) [inline src/core/include/openvino/runtime/tensor.hpp:57]: => preserves

File: src/inference/dev_api/so_ptr.hpp [hunk 3]
```diff
@@ -72,14 +79,14 @@ struct SoPtr {
...
     /**
-     * @brief The shared object or dinamic loaded library
+     * @brief Gets a smart pointer to the custom object
      */
-    std::shared_ptr<void> _so;
+    std::shared_ptr<T> _ptr;
 
     /**
-     * @brief Gets a smart pointer to the custom object
+     * @brief The shared object or dinamic loaded library
```
Comment (ilya-lavrenov) [inline src/inference/dev_api/so_ptr.hpp:87]: dinamic => dynamically

File: src/inference/include/ie/cpp/ie_executable_network.hpp [hunk 4]
```diff
@@ -54,6 +54,11 @@ class INFERENCE_ENGINE_API_CLASS(ExecutableNetwork) {
      */
     ExecutableNetwork() = default;
 
+    /**
+     * @brief Destructor presereves unload order of implementation object and reference to library
```
Comment (ilya-lavrenov) [inline src/inference/include/ie/cpp/ie_executable_network.hpp:58]: presereves => preserves

File: src/inference/include/ie/cpp/ie_infer_request.hpp [hunk 5]
```diff
@@ -66,6 +66,11 @@ class INFERENCE_ENGINE_API_CLASS(InferRequest) {
      */
     InferRequest() = default;
 
+    /**
+     * @brief Destructor presereves unload order of implementation object and reference to library
```
Comment (ilya-lavrenov) [inline src/inference/include/ie/cpp/ie_infer_request.hpp:70]: presereves => preserves

File: src/inference/include/ie/cpp/ie_memory_state.hpp [hunk 6]
```diff
@@ -43,6 +43,11 @@ class INFERENCE_ENGINE_API_CLASS(VariableState) {
      */
     VariableState() = default;
 
+    /**
+     * @brief Destructor presereves unload order of implementation object and reference to library
```
Comment (ilya-lavrenov) [inline src/inference/include/ie/cpp/ie_memory_state.hpp:47]: presereves => preserves

File: src/inference/include/openvino/runtime/compiled_model.hpp [hunk 7]
```diff
@@ -54,6 +54,11 @@ class OPENVINO_RUNTIME_API CompiledModel {
      */
     CompiledModel() = default;
 
+    /**
+     * @brief Destructor presereves unload order of implementation object and reference to library
```
Comment (ilya-lavrenov) [inline src/inference/include/openvino/runtime/compiled_model.hpp:58]: presereves => preserves

File: src/inference/include/openvino/runtime/infer_request.hpp [hunk 8]
```diff
@@ -52,6 +52,13 @@ class OPENVINO_RUNTIME_API InferRequest {
      */
     InferRequest() = default;
 
+    /**
+     * @brief Destructor presereves unload order of implementation object and reference to library
+     * @note To preserve destruction order inside default generated assignment operator we store `_impl` before `_so`.
+     *       And use destructor to remove implementation object before reference to library explicitly
```
Comment (ilya-lavrenov) [inline src/inference/include/openvino/runtime/infer_request.hpp:58]: presereves => preserves

File: src/inference/include/openvino/runtime/remote_context.hpp [hunk 9]
```diff
@@ -37,20 +37,25 @@ class CompiledModel;
...
-     * @param impl Initialized shared pointer
      */
-    RemoteContext(const std::shared_ptr<void>& so, const std::shared_ptr<InferenceEngine::RemoteContext>& impl);
+    RemoteContext(const std::shared_ptr<InferenceEngine::RemoteContext>& impl, const std::shared_ptr<void>& so);
     friend class ov::runtime::Core;
     friend class ov::runtime::CompiledModel;
 
 public:
+    /**
+     * @brief Destructor presereves unload order of implementation object and reference to library
```
Comment (ilya-lavrenov) [inline src/inference/include/openvino/runtime/remote_context.hpp:55]: presereves  => preserves

File: src/inference/include/openvino/runtime/variable_state.hpp [hunk 10]
```diff
@@ -49,6 +49,11 @@ class OPENVINO_RUNTIME_API VariableState {
      */
     VariableState() = default;
 
+    /**
+     * @brief Destructor presereves unload order of implementation object and reference to library
```
Comment (ilya-lavrenov) [inline src/inference/include/openvino/runtime/variable_state.hpp:53]: presereves  => preserves

File: inference-engine/tests/functional/plugin/shared/src/behavior/ov_plugin/life_time.cpp [hunk 11]
```diff
@@ -0,0 +1,105 @@
...
+        objects = {core, compiled_model, request, states, tensor, any, ctx};
+    }
+    for (auto&& i : order) {
+        objects.at(i) = {};
+    }
+}
+
+TEST_P(OVHoldersTest, Orders) {
+    std::vector<std::size_t> order(6);
+    std::iota(order.begin(), order.end(), 0);
```
Comment (mznamens) [inline inference-engine/tests/functional/plugin/shared/src/behavior/ov_plugin/life_time.cpp:63]: Shouldn't it be 1 here as incrementation value?

File: src/core/include/openvino/runtime/allocator.hpp [hunk 12]
```diff
@@ -64,20 +64,25 @@ class Tensor;
...
-     * @param impl Initialized shared pointer
      */
-    Allocator(const std::shared_ptr<void>& so, const AllocatorImpl::Ptr& impl);
+    Allocator(const AllocatorImpl::Ptr& impl, const std::shared_ptr<void>& so);
 
     friend class ov::runtime::Tensor;
 
 public:
+    /**
+     * @brief Destructor presereves unload order of implementation object and reference to library
```
Comment (mznamens) [inline src/core/include/openvino/runtime/allocator.hpp:82]: Typo: preserves

File: src/core/include/openvino/runtime/tensor.hpp [hunk 13]
```diff
@@ -36,23 +36,28 @@ class VariableState;
  */
 class OPENVINO_API Tensor {
 protected:
-    std::shared_ptr<void> _so;                     //!< Reference to dynamicly loaded library
     std::shared_ptr<InferenceEngine::Blob> _impl;  //!< Shared pointer to internal tensor representation
+    std::shared_ptr<void> _so;                     //!< Reference to dynamicly loaded library
```
Comment (mznamens) [inline src/core/include/openvino/runtime/tensor.hpp:40]: Typo - dynamically?

File: src/inference/include/ie/cpp/ie_memory_state.hpp [hunk 14]
```diff
@@ -24,16 +24,16 @@ class IVariableStateInternal;
...
  */
 class INFERENCE_ENGINE_API_CLASS(VariableState) {
-    std::shared_ptr<void> _so;
     std::shared_ptr<IVariableStateInternal> _impl;
+    std::shared_ptr<void> _so;
 
     /**
      * @brief Constructs VariableState from the initialized std::shared_ptr
-     * @param impl Initialized shared pointer
      * @param so Optional: Plugin to use. This is required to ensure that VariableState can work properly even if plugin
```
Comment (mznamens) [inline src/inference/include/ie/cpp/ie_memory_state.hpp:32]: Doxygen comments order is incorrect: should be impl, so

File: src/inference/dev_api/so_ptr.hpp [hunk 15]
```diff
@@ -30,11 +30,18 @@ struct SoPtr {
...
+     * @brief Destructor presereves unload order of implementation object and reference to library
+     */
+    ~SoPtr() {
+        _ptr = {};
+    }
+
     /**
      * @brief Constructs an object with existing shared object reference and loaded pointer
-     * @param soLoader Existing pointer to a library loader
+     * @param so Existing reference to library
```
Comment (mznamens) [inline src/inference/dev_api/so_ptr.hpp:42]: Add ptr description please

**Review by mznamens — COMMENTED** [review]
Nice changes! My comments are minor, except one unclear moment with std::iota and 0-value incrementation

File: inference-engine/tests/functional/plugin/shared/src/behavior/ov_plugin/life_time.cpp (hunk 11 above)
Comment (apankratovantonp) [reply inference-engine/tests/functional/plugin/shared/src/behavior/ov_plugin/life_time.cpp]: It is the first value in range

Comment (mznamens) [reply inference-engine/tests/functional/plugin/shared/src/behavior/ov_plugin/life_time.cpp]: My miss, sorry

**Review by mznamens — APPROVED** [review]

File: src/core/include/openvino/core/any.hpp [hunk 16]
```diff
@@ -342,10 +353,31 @@ class OPENVINO_API Any {
...
 public:
+    /// @brief Default constructor
+    Any() = default;
+
+    /// @brief Default copy constructor
+    /// @param other ohter remote context object
+    Any(const Any& other) = default;
+
+    /// @brief Default copy assignment operator
+    /// @param other ohter remote context object
```
Comment (ilya-lavrenov) [inline src/core/include/openvino/core/any.hpp:364]: ohter remote context object - why remote context?

Comment (apankratovantonp) [reply src/core/include/openvino/core/any.hpp]: fixed

File: src/core/include/openvino/core/any.hpp (hunk 1 above)
Comment (apankratovantonp) [reply src/core/include/openvino/core/any.hpp:378]: done

File: src/inference/include/ie/cpp/ie_infer_request.hpp (hunk 5 above)
Comment (apankratovantonp) [reply src/inference/include/ie/cpp/ie_infer_request.hpp:86]: done

File: src/core/include/openvino/runtime/tensor.hpp [hunk 17]
```diff
@@ -36,35 +36,56 @@ class VariableState;
...
     friend class ov::runtime::InferRequest;
     friend class ov::runtime::RemoteContext;
     friend class ov::runtime::VariableState;
 
 public:
+    /// @brief Default constructor
+    Tensor() = default;
+
+    /// @brief Default copy constructor
+    /// @param other ohter Tensor object
```
Comment (mznamens) [inline src/core/include/openvino/runtime/tensor.hpp:60]: ohter->other here and everywhere

**Review by ilyachur — APPROVED** [review]

**Pull request merged**